### PR TITLE
Elasticsearch upgrading process document

### PR DESCRIPTION
**Pull request by sanjay2916**
Elasticsearch upgrading process document

### CODE REVIEW

File: docs/changelog/0032-upgrade-elasticsearch.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  elasticsearch_download_sha256: 5f7e4bb792917bb7ffc2a5f612dfec87416d54563f795d6a70637befef4cfc6f.
+  ELASTICSEARCH_MAJOR_VERSION=2
+
+  3. update the local settings
+  
+  commcare-cloud <env> update-config
+
+  4.login to any ES server and disable the cluster routing
+  
+  curl -XPUT Elasticsearchserver-IP:9200/_cluster/settings -d '{\"transient\" : {\"cluster.routing.allocation.enable\" : \"none\" }}'
```
Comment (sravfeyn) [inline docs/changelog/0032-upgrade-elasticsearch.md]: This should be persistent not transient. Transient means it will disappear after restart. I have done this already anyway, but good to keep in the docs here.

Comment (sanjay2916) [reply docs/changelog/0032-upgrade-elasticsearch.md]: updated

File: changelog/0032-upgrade-elasticsearch.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
+---
+title: ES upgrade from 1.7.6 to 2.4.6
+key: upgrade-elasticsearch
+date: 2020-02-05
+optional_per_env: yes
+update_steps: |
+  1.stop the site
```
Comment (snopoke) [inline changelog/0032-upgrade-elasticsearch.yml]:
```suggestion
  1. Stop the site
```

File: changelog/0032-upgrade-elasticsearch.yml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  commcare-cloud <env> downtime start
```
Comment (snopoke) [inline changelog/0032-upgrade-elasticsearch.yml]: @sanjay2916 can you put command and config changes inside code blocks

File: changelog/0032-upgrade-elasticsearch.yml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    
+  2.Change the es version and checksum in commcare-cloud/environments/<env>/public.yml
```
Comment (snopoke) [inline changelog/0032-upgrade-elasticsearch.yml]:
```suggestion
  2.Change the ES version and checksum in commcare-cloud/environments/<env>/public.yml 
```

File: changelog/0032-upgrade-elasticsearch.yml [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+update_steps: |
+  1.stop the site
+  commcare-cloud <env> downtime start
+    
+  2.Change the es version and checksum in commcare-cloud/environments/<env>/public.yml 
+  elasticsearch_version: 2.4.6
+  elasticsearch_download_sha256: 5f7e4bb792917bb7ffc2a5f612dfec87416d54563f795d6a70637befef4cfc6f.
+  ELASTICSEARCH_MAJOR_VERSION: 2
+    
+  3.update the local settings
```
Comment (snopoke) [inline changelog/0032-upgrade-elasticsearch.yml]: Markdown requires a space between the `3.` and the text (same comment for all points)
```suggestion
  3. Update the local settings
```

File: changelog/0032-upgrade-elasticsearch.yml (hunk 3 above)
Comment (sanjay2916) [reply changelog/0032-upgrade-elasticsearch.yml]: sure

File: changelog/0032-upgrade-elasticsearch.yml [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  }
+    
+  3. Update the local settings
+  ```bash
+  commcare-cloud <env> update-config
+  ```  
+  4. Login to any ES server and disable the cluster routing
+  ```bash
+  curl -XPUT Elasticsearchserver-IP:9200/_cluster/settings -d '{\"persistent\" : {\"cluster.routing.allocation.enable\" : \"none\" }}'
+  ```
```
[the remark this answers is not in the document]
Comment (esoergel) [reply changelog/0032-upgrade-elasticsearch.yml]: I see now one of the commands below uses `<ES-IP>`, which looks good.  This should be consistent throughout the instructions.

File: changelog/0032-upgrade-elasticsearch.yml [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    
+  11. Verify that the cluster status is yellow
+  ```bash
+  curl -XGET <ES-IP>:9200/_cluster/health?pretty
+  ```
+    
+  12. Enable the cluster routing
+  ```bash
+  curl -XPUT <ES-IP>:9200/_cluster/settings -d '{\"persistent\" : {\"cluster.routing.allocation.enable\" : \"all\" }}'
```
Comment (sravfeyn) [inline changelog/0032-upgrade-elasticsearch.yml]: Btw, this shouldn't be done on ICDS @sanjay2916 (no need to update docs, since this is for all envs)

File: changelog/0032-upgrade-elasticsearch.yml [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+  ```bash
+  commcare-cloud <env> run-shell-command all "service monit stop" -b --limit=elasticsearch
+  commcare-cloud <env> run-shell-command all "service elasticsearch stop" -b --limit=elasticsearch
+  ```
+    
+  8. Rename the 1.7.6 ES data folder with new version 2.4.6
+  ```bash
+  commcare-cloud <env> run-shell-command all "mv /opt/data/elasticsearch-2.4.6 /opt/data/elasticsearch-2.4.6-new-installation" -b --limit=elasticsearch
+  commcare-cloud <env> run-shell-command all "mv /opt/data/elasticsearch-1.7.6 /opt/data/elasticsearch-2.4.6" -b --limit=elasticsearch
+  ```
```
[the remark this answers is not in the document]
Comment (dannyroberts) [reply changelog/0032-upgrade-elasticsearch.yml]: @sanjay2916 Shouldn't we be telling people to first copy the data and _then_ install the new version of Elasticsearch (move this step to before step 6)? Then we wouldn't need step 7, and we wouldn't need to `mv /opt/data/elasticsearch-2.4.6 /opt/data/elasticsearch-2.4.6-new-installation` because `/opt/data/elasticsearch-2.4.6` wouldn't exist yet.

File: changelog/0032-upgrade-elasticsearch.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (dannyroberts) [reply changelog/0032-upgrade-elasticsearch.yml]: Throughout instead of `<ES-IP>`, you can use `$(cchq ${ENV} lookup elasticsearch:0)` which will actually calculate that. Since there are so many commands to run, I might also tell them to set their ENV environment variable, and then use `${ENV}` throughout these commands

[the remark this answers is not in the document]
Comment (dannyroberts) [reply changelog/0032-upgrade-elasticsearch.yml]: https://www.elastic.co/guide/en/elasticsearch/reference/2.4/restart-upgrade.html#_step_1_disable_shard_allocation_2 also suggests performing a `POST /_flush/synced` before shutting down.

File: changelog/0032-upgrade-elasticsearch.yml [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+  ELASTICSEARCH_MAJOR_VERSION: 2
+  ```
+    
+  3. Update the local settings
+  ```bash
+  commcare-cloud <env> update-config
+  ```  
+  4. Disable the cluster routing
+  ```bash
+  curl -XPUT <ES-IP>:9200/_cluster/settings -d '{\"persistent\" : {\"cluster.routing.allocation.enable\" : \"none\" }}'
```
Comment (dannyroberts) [inline changelog/0032-upgrade-elasticsearch.yml]: nit: I don't think you need to escape the `"`s here with a `\` since they're wrapped with `'`. You also might be required to have the content type application/json header, though I'm not 100% sure. I'd suggest you go to https://www.elastic.co/guide/en/elasticsearch/reference/2.4/restart-upgrade.html and click on "Copy as cURL" for each of these curl commands, and then replace "localsettings" with `$(cchq ${ENV} lookup elasticsearch:0)`

File: changelog/0032-upgrade-elasticsearch.yml [hunk 10]
```diff
@@ -0,0 +1,84 @@
+---
+title: ES upgrade from 1.7.6 to 2.4.6
+key: upgrade-elasticsearch
+date: 2020-02-05
+optional_per_env: no
+min_commcare_version: 19be50699bf78b25a8da6e3e13eb5bc6d13245e8
```
Comment (dannyroberts) [inline changelog/0032-upgrade-elasticsearch.yml]: I think this is the wrong commit, looks like it was copied from `0030-upgrade-sentry.yml`. You can use 064bb6dfc2b333cc8e6e3955b0c5d30513c5f4be instead.

File: changelog/0032-upgrade-elasticsearch.yml (hunk 9 above)
Comment (dannyroberts) [reply changelog/0032-upgrade-elasticsearch.yml]: Throughout instead of <ES-IP>, you can use $(cchq ${ENV} lookup elasticsearch:0) which will actually calculate that. Since there are so many commands to run, I might also tell them to set their ENV environment variable, and then use ${ENV} throughout these commands

Comment (dannyroberts) [reply changelog/0032-upgrade-elasticsearch.yml]: https://www.elastic.co/guide/en/elasticsearch/reference/2.4/restart-upgrade.html#_step_1_disable_shard_allocation_2 also suggests performing a `POST /_flush/synced` before shutting down.

File: changelog/0032-upgrade-elasticsearch.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (sanjay2916) [reply changelog/0032-upgrade-elasticsearch.yml]: @dannyroberts  After we run deploy stack , the es 2.4.6 empty directory will create in /opt/data/  folder and if you move 1.7.6 to 2.4.6 before renaming of 2.4.6 directory it will go inside the 2.4.6 directory

File: changelog/0032-upgrade-elasticsearch.yml [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+    "persistent": {
+       "cluster.routing.allocation.enable": "none"
+    }
+  }
+  '
+  ```
+
+  5. Perform a synced flush
+  ```bash
+  curl -X POST ""<ES-IP>:9200/_flush/synced?pretty"
```
Comment (esoergel) [inline changelog/0032-upgrade-elasticsearch.yml]:
```suggestion
  curl -X POST "<ES-IP>:9200/_flush/synced?pretty"
```

File: docs/commcare-cloud/commands/index.md [hunk 12]
```diff
@@ -566,8 +566,7 @@ tmux windows will a new one be opened.
 Export Sentry events. One line per event JSON.
 
 ```
-commcare-cloud <env> export-sentry-events -k API_KEY -p PROJECT_ID [-q QUERY] [--start START] [--end END]
-                                          [--organization ORGANIZATION] [--full]
+commcare-cloud <env> export-sentry-events -k API_KEY -i ISSUE_ID [--full] [--cursor CURSOR]
```
Comment (esoergel) [inline docs/commcare-cloud/commands/index.md]: Were the changes in this file committed by mistake?  They seem unrelated to the rest of the PR.

Comment (dannyroberts) [reply docs/commcare-cloud/commands/index.md]: I'm guessing this has to do with an external change from another PR? If the build is failing it means that the files match what `make` makes

Comment (dannyroberts) [reply docs/commcare-cloud/commands/index.md]: I'll merge and then test it locally to make sure, and make a PR with any touch ups necessary

File: changelog/0032-upgrade-elasticsearch.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (dannyroberts) [reply changelog/0032-upgrade-elasticsearch.yml]: Hey @sanjay2916, I was asking something different. I was asking, why not use the following steps?

1. stop elasticsearch
2. `cp -r /opt/data/elasticsearch-1.7.6 /opt/data/elasticsearch-2.4.6`
3. Verify permissions and size of data directory
4. Install new elasticsearch with `commcare-cloud <env> ansible-playbook deploy_stack.yml --limit=elasticsearch --tags=elasticsearch` (which also starts elasticsearch automatically)
5. Verify the cluster status is yellow
6. ...

Was there some reason to move the data _after_ updating ES instead of before it?

Regardless, since you've tested it many times the way you have it here, maybe it's best to keep it this way rather than asking people to do something that's not well-tested.
